pkg/frontend/middleware/log.go

```diff
@@ -96,16 +99,75 @@ func Log(baseLog *logrus.Entry) func(http.Handler) http.Handler {
 			})
 			log.Print("read request")
 
+			var (
+				auditCallerIdentity                              = r.UserAgent()
+				auditCallerType                                  = audit.CallerIdentityTypeApplicationID
+				auditTargetResourceName, auditTargetResourceType = auditTargetResourceData(r)
+			)
+
+			if correlationData.ClientPrincipalName != "" {
+				auditCallerIdentity = correlationData.ClientPrincipalName
+				auditCallerType = audit.CallerIdentityTypeObjectID
+			}
+
+			auditEntry := auditLog.WithFields(logrus.Fields{
+				audit.EnvKeyAppID:             audit.SourceRP,
+				audit.EnvKeyCloudRole:         audit.CloudRoleRP,
+				audit.EnvKeyCorrelationID:     correlationData.CorrelationID,
+				audit.EnvKeyEnvironment:       env.Environment().Name,
+				audit.EnvKeyHostname:          env.Hostname(),
+				audit.EnvKeyLocation:          env.Location(),
+				audit.PayloadKeyCategory:      audit.CategoryResourceManagement,
+				audit.PayloadKeyOperationName: fmt.Sprintf("%s %s", r.Method, r.URL.Path),
+				audit.PayloadKeyRequestID:     correlationData.RequestID,
+				audit.PayloadKeyCallerIdentities: []audit.CallerIdentity{
+					{
+						CallerIdentityType:  auditCallerType,
+						CallerIdentityValue: auditCallerIdentity,
+						CallerIPAddress:     r.RemoteAddr,
+					},
+				},
+				audit.PayloadKeyTargetResources: []audit.TargetResource{
+					{
+						TargetResourceName: auditTargetResourceName,
+						TargetResourceType: auditTargetResourceType,
+					},
+				},
+			})
+
 			defer func() {
+				statusCode := w.(*logResponseWriter).statusCode
 				log.WithFields(logrus.Fields{
 					"body_read_bytes":      r.Body.(*logReadCloser).bytes,
 					"body_written_bytes":   w.(*logResponseWriter).bytes,
 					"duration":             time.Since(t).Seconds(),
-					"response_status_code": w.(*logResponseWriter).statusCode,
+					"response_status_code": statusCode,
 				}).Print("sent response")
+
+				resultType := audit.ResultTypeSuccess
+				if statusCode >= http.StatusBadRequest {
+					resultType = audit.ResultTypeFail
+				}
+
+				auditEntry.WithFields(logrus.Fields{
+					audit.PayloadKeyResult: audit.Result{
+						ResultType:        resultType,
+						ResultDescription: fmt.Sprintf("Status code: %d", statusCode),
+					},
+				}).Info(audit.DefaultLogMessage)
 			}()
 
 			h.ServeHTTP(w, r)
 		})
 	}
 }
+
+func auditTargetResourceData(r *http.Request) (string, string) {
```

Comment:
Do you think it is worth adding simple unit test for this function to make sure all our current "URL patterns" will return what we want them to return?